modules/navier_stokes/src/fvkernels/PINSFVMomentumFrictionCorrection.C

```diff
@@ -0,0 +1,132 @@
+
+InputParameters
+PINSFVMomentumFrictionCorrection::validParams()
+{
+  auto params = INSFVFluxKernel::validParams();
+  params.addClassDescription(
+      "Computes the diffusion kernel to avoid pressure-driven oscillations.");
+  params.addParam<MooseFunctorName>("Darcy_name", "Name of the Darcy coefficients property.");
+  params.addParam<MooseFunctorName>("Forchheimer_name",
+                                    "Name of the Forchheimer coefficients property.");
+  params.addParam<MooseFunctorName>(NS::porosity, NS::porosity, "the porosity");
+  params.addRequiredParam<MooseFunctorName>(NS::density, "The density.");
+  params.addRangeCheckedParam<Real>("consistent_scaling",
+                                    1,
+                                    "consistent_scaling >= 0",
+                                    "Smoothing scaling parameter to control "
+                                    "collocated mesh oscillations");
+  return params;
+}
+
+PINSFVMomentumFrictionCorrection::PINSFVMomentumFrictionCorrection(const InputParameters & params)
+  : INSFVFluxKernel(params),
+    _cL(isParamValid("Darcy_name") ? &getFunctor<ADRealVectorValue>("Darcy_name") : nullptr),
+    _cQ(isParamValid("Forchheimer_name") ? &getFunctor<ADRealVectorValue>("Forchheimer_name")
+                                         : nullptr),
+    _use_Darcy_friction_model(isParamValid("Darcy_name")),
+    _use_Forchheimer_friction_model(isParamValid("Forchheimer_name")),
+    _eps(getFunctor<ADReal>(NS::porosity)),
+    _rho(getFunctor<ADReal>(NS::density)),
+    _consistent_scaling(getParam<Real>("consistent_scaling"))
+{
+  if (!_use_Darcy_friction_model && !_use_Forchheimer_friction_model)
+    mooseError("At least one friction model needs to be specified.");
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError("PINSFV is not supported by local AD indexing. In order to use PINSFV, please run "
+             "the configure script in the root MOOSE directory with the configure option "
+             "'--with-ad-indexing-type=global'");
+#endif
+}
+
+void
+PINSFVMomentumFrictionCorrection::gatherRCData(const FaceInfo & fi)
+{
+  if (skipForBoundary(fi))
+    return;
+
+  _face_info = &fi;
+  _normal = fi.normal();
+  _face_type = fi.faceType(_var.name());
+
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+  using namespace Moose::FV;
+
+  const auto elem_face = elemFromFace();
+  const auto neighbor_face = neighborFromFace();
+
+  ADReal friction_term_elem = 0;
+  ADReal friction_term_neighbor = 0;
+
+  if (_use_Darcy_friction_model)
+  {
+    friction_term_elem += (*_cL)(elem_face)(_index)*_rho(elem_face) / _eps(elem_face);
+    friction_term_neighbor +=
+        (*_cL)(neighbor_face)(_index)*_rho(neighbor_face) / _eps(neighbor_face);
+  }
+  if (_use_Forchheimer_friction_model)
+  {
+    friction_term_elem += (*_cQ)(elem_face)(_index)*_rho(elem_face) / _eps(elem_face);
+    friction_term_neighbor +=
+        (*_cQ)(neighbor_face)(_index)*_rho(neighbor_face) / _eps(neighbor_face);
+  }
+
```

Comment:
The derivation for this kernel works out to use the *coefficients* that multiply the dofs in `PINSFVMomentumFriction`. The 'B' data held by the interpolator UO corresponds to the coefficient times dof product.

Reply:
We can always divide the coefs of `PINSFVMomentumFriction` with the current dofs, but it is risky because we may divide by zero. I think the current way is better although it introduces a bit of an extra computation effort. Optimally, we should have a separate user object to compute only the coefficients and then use it to construct the body force and friction correction kernels. What do you think @lindsayad @GiudGiud ?

Reply:
We would have to introduce a distinction between body forces that are functions of velocity, like friction, and body forces that are not, such as gravity. This is probably a good idea for saving compute expense because we could handle the residuals from 
1. body force terms that are functions of the velocity
2. their correction counterparts

in a single place, but it would require more work. I can do it here or in a follow-on

Reply:
The new interpolator API that objects like `PINSFVMomentumFriction` would use would be something like `addToF` in the vein of the notation that I've seen @tanoret use

Reply:
Makes sense. That would be useful